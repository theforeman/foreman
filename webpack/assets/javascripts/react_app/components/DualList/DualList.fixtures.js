export const id = 'some-id';

export const selectedItems = [
  { label: 'CentOS', value: 1 },
  { label: 'Fedora', value: 2 },
];

export const props = {
  items: [
    { label: 'CentOS', value: 1 },
    { label: 'Fedora', value: 2 },
    { label: 'RedHat', value: 3 },
    { label: 'Debian', value: 4 },
    { label: 'Ubuntu', value: 5 },
  ],
  selectedIDs: [1, 2],
  inputProps: { name: 'some-name', id: 'some-id' },
  label: 'some -label',
  id,
};

export const initialData = {
  id,
  selectedItems,
};

export const itemsChanged = {
  id,
  selectedItems,
};
