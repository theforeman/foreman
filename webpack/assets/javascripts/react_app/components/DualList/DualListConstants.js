export const DUAL_LIST_INIT = 'DUAL_LIST_INIT';
export const DUAL_LIST_CHANGE = 'DUAL_LIST_CHANGE';
