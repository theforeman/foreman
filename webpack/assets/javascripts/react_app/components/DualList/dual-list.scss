@import '~patternfly/dist/sass/patternfly/_color-variables.scss';
@import '~patternfly-react/dist/sass/_dual-list-selector.scss';

.dual_list_error {
  color: red;
}
