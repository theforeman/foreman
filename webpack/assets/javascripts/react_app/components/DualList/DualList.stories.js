import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs } from '@storybook/addon-knobs';
import { Col } from 'patternfly-react';
import DualList from './DualList';
import { props } from './DualList.fixtures';
import Story from '../../../../../stories/components/Story';

storiesOf('Components/DualList', module)
  .addDecorator(withKnobs)
  .add('DualList', () => (
    <Story>
      <Col md={8} mdOffset={2}>
        <DualList {...props} />
      </Col>
    </Story>
  ));
