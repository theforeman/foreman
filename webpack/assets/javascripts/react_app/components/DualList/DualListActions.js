import { DUAL_LIST_INIT, DUAL_LIST_CHANGE } from './DualListConstants';

export const initialUpdate = (selectedItems, id) => ({
  type: DUAL_LIST_INIT,
  payload: {
    selectedItems,
    id,
  },
});

export const onChange = (selectedItems, id) => ({
  type: DUAL_LIST_CHANGE,
  payload: {
    selectedItems,
    id,
  },
});
