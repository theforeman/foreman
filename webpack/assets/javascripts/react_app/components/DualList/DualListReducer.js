import Immutable from 'seamless-immutable';
import { DUAL_LIST_INIT, DUAL_LIST_CHANGE } from './DualListConstants';

export default (state = Immutable({}), action) => {
  const { type, payload: { selectedItems, id } = {} } = action;
  switch (type) {
    case DUAL_LIST_INIT:
      return state.setIn([id], {
        selectedItems,
      });
    case DUAL_LIST_CHANGE:
      return state.setIn([id], {
        selectedItems,
      });
    default:
      return state;
  }
};
