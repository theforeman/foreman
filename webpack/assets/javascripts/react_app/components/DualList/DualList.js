import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { DualListControlled as PfDualList } from 'patternfly-react';
import { arrangeItemsBySelectedIDs } from './helpers';
import { bindMethods, noop } from '../../common/helpers';
import FormField from '../common/forms/FormField';
import './dual-list.scss';

class DualList extends React.Component {
  constructor(props) {
    super(props);
    bindMethods(this, ['handleInit', 'handleChange']);
  }

  handleInit({ right: { items } }) {
    const { initialUpdate, id } = this.props;
    const selectedItems = items.map(({ label, value }) => ({ label, value }));
    initialUpdate(selectedItems, id);
  }

  handleChange({ right: { items } }) {
    const { onChange, id } = this.props;
    const selectedItems = items.map(({ label, value }) => ({ label, value }));
    onChange(selectedItems, id);
  }

  render() {
    const { inputProps, items, selectedIDs, label, id, error } = this.props;
    const { selectedList, unselectedlist } = arrangeItemsBySelectedIDs(
      items,
      selectedIDs
    );
    return (
      <div id={id}>
        <FormField
          labelSizeClass="col-md-2"
          inputSizeClass="col-md-10"
          label={label}
        >
          <Fragment>
            <PfDualList
              allowHiddenInputs
              onComponentInit={this.handleInit}
              onChange={this.handleChange}
              left={{
                items: unselectedlist,
                inputProps: {
                  className: inputProps.className,
                },
              }}
              right={{
                items: selectedList,
                inputProps,
              }}
            />
            <div className="dual_list_error">{error}</div>
          </Fragment>
        </FormField>
      </div>
    );
  }
}

DualList.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  error: PropTypes.string,
  inputProps: PropTypes.object,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
        .isRequired,
    })
  ),
  selectedIDs: PropTypes.arrayOf(
    PropTypes.oneOfType([PropTypes.string, PropTypes.number])
  ),
  initialUpdate: PropTypes.func,
  onChange: PropTypes.func,
};

DualList.defaultProps = {
  error: null,
  inputProps: {},
  items: [],
  selectedIDs: [],
  initialUpdate: noop,
  onChange: noop,
};

export default DualList;
