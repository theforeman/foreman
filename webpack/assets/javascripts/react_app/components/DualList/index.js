import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as actions from './DualListActions';
import reducer from './DualListReducer';
import DualList from './DualList';

const mapStateToProps = (state, { id }) => ({});

const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export const reducers = { dualList: reducer };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DualList);
