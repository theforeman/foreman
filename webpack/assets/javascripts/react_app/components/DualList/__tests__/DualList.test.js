import DualList from '../DualList';
import { props } from '../DualList.fixtures';
import { testComponentSnapshotsWithFixtures } from '../../../common/testHelpers';

const fixtures = {
  'renders DualList': props,
};
describe('DualList', () => {
  describe('rendering', () => {
    testComponentSnapshotsWithFixtures(DualList, fixtures);
  });
});
