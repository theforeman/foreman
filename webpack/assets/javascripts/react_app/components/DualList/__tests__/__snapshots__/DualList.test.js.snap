// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DualList rendering renders DualList 1`] = `
<div
  id="some-id"
>
  <FormField
    className=""
    disabled={false}
    helpInline={null}
    id={null}
    inputProps={null}
    inputSizeClass="col-md-10"
    label="some -label"
    labelHelp={null}
    labelSizeClass="col-md-2"
    onChange={[Function]}
    required={false}
    touched={false}
    type="text"
  >
    <DualListControlled
      allowHiddenInputs={true}
      left={
        Object {
          "inputProps": Object {
            "className": undefined,
          },
          "items": Array [
            Object {
              "label": "RedHat",
              "value": 3,
            },
            Object {
              "label": "Debian",
              "value": 4,
            },
            Object {
              "label": "Ubuntu",
              "value": 5,
            },
          ],
        }
      }
      onChange={[Function]}
      onComponentInit={[Function]}
      onFilterChange={[Function]}
      onItemChange={[Function]}
      onMainCheckboxChange={[Function]}
      onSortClick={[Function]}
      right={
        Object {
          "inputProps": Object {
            "id": "some-id",
            "name": "some-name",
          },
          "items": Array [
            Object {
              "label": "CentOS",
              "value": 1,
            },
            Object {
              "label": "Fedora",
              "value": 2,
            },
          ],
        }
      }
    />
    <div
      className="dual_list_error"
    />
  </FormField>
</div>
`;
